File: server/internal/models/recommendation.go
package models

import "time"

// Recommendation type values.
const (
	RecTypeBook    = "book"
	RecTypeMovie   = "movie"
	RecTypeTV      = "tv"
	RecTypeMusic   = "music"
	RecTypeArticle = "article"
	RecTypePodcast = "podcast"
	RecTypeOther   = "other"
)

// Recommendation status values.
const (
	RecStatusNew       = "new"
	RecStatusSaved     = "saved"
	RecStatusDone      = "done"
	RecStatusDismissed = "dismissed"
)

// Recommendation is the list-item representation.
type Recommendation struct {
	ID                   string    `json:"id"`
	Type                 string    `json:"type"`
	Title                string    `json:"title"`
	Creator              string    `json:"creator,omitempty"`
	SourceNewsletterName string    `json:"source_newsletter_name"`
	SourceEmailID        string    `json:"source_email_id,omitempty"`
	SourceDate           time.Time `json:"source_date"`
	ContextSnippet       string    `json:"context_snippet"`
	Status               string    `json:"status"`
	DuplicateCount       int       `json:"duplicate_count"`
	IsUserAdded          bool      `json:"is_user_added"`
	CreatedAt            time.Time `json:"created_at"`
}

// RecommendationDetail includes full context and duplicate sources.
type RecommendationDetail struct {
	Recommendation   Recommendation    `json:"recommendation"`
	FullContext      string            `json:"full_context,omitempty"`
	DuplicateSources []DuplicateSource `json:"duplicate_sources"`
}

// DuplicateSource tracks where a recommendation was mentioned.
type DuplicateSource struct {
	NewsletterName string    `json:"newsletter_name"`
	EmailID        string    `json:"email_id,omitempty"`
	Date           time.Time `json:"date"`
	ContextSnippet string    `json:"context_snippet"`
}

// RecommendationCreateRequest is the request body for user-added recommendations.
type RecommendationCreateRequest struct {
	Type           string `json:"type"`
	Title          string `json:"title"`
	Creator        string `json:"creator,omitempty"`
	ContextSnippet string `json:"context_snippet,omitempty"`
}

// RecommendationUpdateRequest is the request body for status changes.
type RecommendationUpdateRequest struct {
	Status string `json:"status"`
}

// RecommendationListResponse is the paginated response for recommendation listings.
type RecommendationListResponse struct {
	Data       []Recommendation `json:"data"`
	NextCursor string           `json:"next_cursor,omitempty"`
	HasMore    bool             `json:"has_more"`
}

// ValidRecommendationTypes returns all valid recommendation type values.
func ValidRecommendationTypes() []string {
	return []string{RecTypeBook, RecTypeMovie, RecTypeTV, RecTypeMusic, RecTypeArticle, RecTypePodcast, RecTypeOther}
}

// ValidRecommendationStatuses returns all valid recommendation status values.
func ValidRecommendationStatuses() []string {
	return []string{RecStatusNew, RecStatusSaved, RecStatusDone, RecStatusDismissed}
}

File: clients/shared/Sources/EmailClientKit/Models/Recommendation.swift
import Foundation

/// The type/category of a recommendation.
public enum RecommendationType: String, Codable, Sendable, Equatable, CaseIterable {
    case book
    case movie
    case tv
    case music
    case article
    case podcast
    case other
}

/// User-managed status of a recommendation.
public enum RecommendationStatus: String, Codable, Sendable, Equatable, CaseIterable {
    case new
    case saved
    case done
    case dismissed
}

/// A recommendation extracted from a newsletter.
public struct Recommendation: Codable, Identifiable, Sendable, Equatable {
    public let id: String
    public let type: RecommendationType
    public let title: String
    public let creator: String?
    public let sourceNewsletterName: String
    public let sourceEmailId: String?
    public let sourceDate: Date
    public let contextSnippet: String
    public let status: RecommendationStatus
    public let duplicateCount: Int
    public let isUserAdded: Bool?
    public let createdAt: Date?

    public init(
        id: String,
        type: RecommendationType,
        title: String,
        creator: String? = nil,
        sourceNewsletterName: String,
        sourceEmailId: String? = nil,
        sourceDate: Date,
        contextSnippet: String,
        status: RecommendationStatus,
        duplicateCount: Int,
        isUserAdded: Bool? = nil,
        createdAt: Date? = nil
    ) {
        self.id = id
        self.type = type
        self.title = title
        self.creator = creator
        self.sourceNewsletterName = sourceNewsletterName
        self.sourceEmailId = sourceEmailId
        self.sourceDate = sourceDate
        self.contextSnippet = contextSnippet
        self.status = status
        self.duplicateCount = duplicateCount
        self.isUserAdded = isUserAdded
        self.createdAt = createdAt
    }
}

/// Detail view of a recommendation with duplicate sources.
public struct RecommendationDetail: Codable, Sendable, Equatable {
    public let recommendation: Recommendation
    public let fullContext: String?
    public let duplicateSources: [DuplicateSource]

    public init(
        recommendation: Recommendation,
        fullContext: String? = nil,
        duplicateSources: [DuplicateSource]
    ) {
        self.recommendation = recommendation
        self.fullContext = fullContext
        self.duplicateSources = duplicateSources
    }
}

/// A source that recommended the same item.
public struct DuplicateSource: Codable, Sendable, Equatable {
    public let newsletterName: String
    public let emailId: String?
    public let date: Date
    public let contextSnippet: String

    public init(
        newsletterName: String,
        emailId: String? = nil,
        date: Date,
        contextSnippet: String
    ) {
        self.newsletterName = newsletterName
        self.emailId = emailId
        self.date = date
        self.contextSnippet = contextSnippet
    }
}
